File: ios/MapTalk/Data/FirestoreSchema.swift
import FirebaseFirestore
import Foundation

/// Collection and field names, mirrored in
/// `android/app/src/main/java/app/maptalk/data/FirestoreSchema.kt` and enforced by
/// `firebase/firestore.rules`.
enum Fs {
    static let threads = "threads"
    static let messages = "messages"
    static let users = "users"

    static let title = "title"
    static let kind = "kind"
    static let lat = "lat"
    static let lng = "lng"
    static let geohash = "geohash"
    static let authorId = "authorId"
    static let authorName = "authorName"
    static let createdAt = "createdAt"
    static let lastMessageAt = "lastMessageAt"
    static let messageCount = "messageCount"
    static let lastMediaPath = "lastMediaPath"
    static let lastMediaKind = "lastMediaKind"
    static let text = "text"
    static let displayName = "displayName"
    static let photoURL = "photoURL"
    static let photoPath = "photoPath"
    static let updatedAt = "updatedAt"
    static let kindMessage = "messageKind"
    static let imagePath = "imagePath"
    static let imageWidth = "imageWidth"
    static let imageHeight = "imageHeight"
    static let audioPath = "audioPath"
    static let audioDurationMs = "audioDurationMs"
    static let videoPath = "videoPath"
    static let videoDurationMs = "videoDurationMs"
    static let videoWidth = "videoWidth"
    static let videoHeight = "videoHeight"
    static let replyToId = "replyToId"
    static let replyToText = "replyToText"
    static let replyToAuthorName = "replyToAuthorName"
    static let reactions = "reactions"

    static let blocks = "blocks"
    static let blockedUid = "blockedUid"
    static let reports = "reports"
    static let targetType = "targetType"
    static let targetId = "targetId"
    static let threadId = "threadId"
    static let targetAuthorId = "targetAuthorId"
    static let reason = "reason"

    static let devices = "devices"
    static let token = "token"
    static let platform = "platform"
    static let subscribers = "subscribers"
    static let subscribedAt = "subscribedAt"
}

extension DocumentSnapshot {

    func chatThread() -> ChatThread? {
        guard let title = self[Fs.title] as? String,
              let lat = self[Fs.lat] as? Double,
              let lng = self[Fs.lng] as? Double
        else { return nil }

        return ChatThread(
            id: documentID,
            title: title,
            kind: ThreadKind(id: self[Fs.kind] as? String),
            position: GeoPoint(lat: lat, lng: lng),
            geohash: self[Fs.geohash] as? String ?? "",
            authorId: self[Fs.authorId] as? String ?? "",
            authorName: self[Fs.authorName] as? String ?? "",
            createdAt: estimatedDate(Fs.createdAt),
            lastMessageAt: estimatedDate(Fs.lastMessageAt),
            messageCount: (self[Fs.messageCount] as? NSNumber)?.intValue ?? 0,
            lastMediaPath: self[Fs.lastMediaPath] as? String,
            lastMediaKind: (self[Fs.lastMediaKind] as? String).flatMap(MessageKind.init(rawValue:))
        )
    }

    func message() -> Message? {
        let text = self[Fs.text] as? String ?? ""
        let kind = MessageKind(rawValue: self[Fs.kindMessage] as? String ?? "") ?? .text
        switch kind {
        case .text where text.isEmpty: return nil
        case .image where self[Fs.imagePath] as? String == nil: return nil
        case .voice where self[Fs.audioPath] as? String == nil: return nil
        case .video where self[Fs.videoPath] as? String == nil: return nil
        case .sticker where text.isEmpty: return nil
        default: break
        }

        var reply: MessageReply?
        if let replyId = self[Fs.replyToId] as? String,
           let replyName = self[Fs.replyToAuthorName] as? String,
           let replyText = self[Fs.replyToText] as? String {
            reply = MessageReply(id: replyId, authorName: replyName, text: replyText)
        }

        var reactions: [String: [String]] = [:]
        if let raw = self[Fs.reactions] as? [String: Any] {
            for (emoji, value) in raw {
                if let uids = value as? [String] {
                    reactions[emoji] = uids
                } else if let uids = value as? [Any] {
                    reactions[emoji] = uids.compactMap { $0 as? String }
                }
            }
        }

        return Message(
            id: documentID,
            kind: kind,
            text: text,
            authorId: self[Fs.authorId] as? String ?? "",
            authorName: self[Fs.authorName] as? String ?? "",
            createdAt: estimatedDate(Fs.createdAt),
            imagePath: self[Fs.imagePath] as? String,
            imageWidth: (self[Fs.imageWidth] as? NSNumber)?.intValue,
            imageHeight: (self[Fs.imageHeight] as? NSNumber)?.intValue,
            audioPath: self[Fs.audioPath] as? String,
            audioDurationMs: (self[Fs.audioDurationMs] as? NSNumber)?.intValue,
            videoPath: self[Fs.videoPath] as? String,
            videoDurationMs: (self[Fs.videoDurationMs] as? NSNumber)?.intValue,
            videoWidth: (self[Fs.videoWidth] as? NSNumber)?.intValue,
            videoHeight: (self[Fs.videoHeight] as? NSNumber)?.intValue,
            reply: reply,
            reactions: reactions
        )
    }

    private func estimatedDate(_ field: String) -> Date? {
        (get(field, serverTimestampBehavior: .estimate) as? Timestamp)?.dateValue()
    }
}

File: android/app/src/main/java/app/maptalk/data/FirestoreSchema.kt
package app.maptalk.data

import app.maptalk.data.model.ChatThread
import app.maptalk.data.model.Message
import app.maptalk.data.model.MessageKind
import app.maptalk.data.model.MessageReply
import app.maptalk.data.model.ThreadKind
import app.maptalk.geo.GeoPoint
import com.google.firebase.Timestamp
import com.google.firebase.firestore.DocumentSnapshot
import java.time.Instant

/**
 * Collection and field names, mirrored in `ios/MapTalk/Data/FirestoreSchema.swift` and
 * enforced by `firebase/firestore.rules`.
 */
object Fs {
    const val THREADS = "threads"
    const val MESSAGES = "messages"
    const val USERS = "users"

    const val TITLE = "title"
    const val KIND = "kind"
    const val LAT = "lat"
    const val LNG = "lng"
    const val GEOHASH = "geohash"
    const val AUTHOR_ID = "authorId"
    const val AUTHOR_NAME = "authorName"
    const val AUTHOR_PHOTO_URL = "authorPhotoURL"
    const val CREATED_AT = "createdAt"
    const val LAST_MESSAGE_AT = "lastMessageAt"
    const val MESSAGE_COUNT = "messageCount"
    const val LAST_MEDIA_PATH = "lastMediaPath"
    const val LAST_MEDIA_KIND = "lastMediaKind"
    const val TEXT = "text"
    const val DISPLAY_NAME = "displayName"
    const val PHOTO_URL = "photoURL"
    const val PHOTO_PATH = "photoPath"
    const val MESSAGE_KIND = "messageKind"
    const val IMAGE_PATH = "imagePath"
    const val IMAGE_WIDTH = "imageWidth"
    const val IMAGE_HEIGHT = "imageHeight"
    const val AUDIO_PATH = "audioPath"
    const val AUDIO_DURATION_MS = "audioDurationMs"
    const val VIDEO_PATH = "videoPath"
    const val VIDEO_DURATION_MS = "videoDurationMs"
    const val VIDEO_WIDTH = "videoWidth"
    const val VIDEO_HEIGHT = "videoHeight"
    const val REPLY_TO_ID = "replyToId"
    const val REPLY_TO_TEXT = "replyToText"
    const val REPLY_TO_AUTHOR_NAME = "replyToAuthorName"
    const val REACTIONS = "reactions"
    const val EDITED_AT = "editedAt"

    const val BLOCKS = "blocks"
    const val BLOCKED_UID = "blockedUid"
    const val REPORTS = "reports"
    const val TARGET_TYPE = "targetType"
    const val TARGET_ID = "targetId"
    const val THREAD_ID = "threadId"
    const val TARGET_AUTHOR_ID = "targetAuthorId"
    const val REASON = "reason"

    const val DEVICES = "devices"
    const val TOKEN = "token"
    const val PLATFORM = "platform"
    const val UPDATED_AT = "updatedAt"
    const val SUBSCRIBERS = "subscribers"
    const val SUBSCRIBED_AT = "subscribedAt"

    /** Precision used for the stored hash; the query bounds use their own precision. */
    const val GEOHASH_PRECISION = 10
}

/**
 * Documents are mapped by hand rather than through reflection, so a schema change is a
 * compile error instead of a silent null, and R8 needs no keep rules.
 *
 * A locally created document has no server timestamp yet, so timestamps are read with
 * [DocumentSnapshot.ServerTimestampBehavior.ESTIMATE]. That gives just-sent messages a
 * usable sort key instead of null.
 */
fun DocumentSnapshot.toChatThread(): ChatThread? {
    val title = getString(Fs.TITLE) ?: return null
    val lat = getDouble(Fs.LAT) ?: return null
    val lng = getDouble(Fs.LNG) ?: return null
    return ChatThread(
        id = id,
        title = title,
        kind = ThreadKind.fromId(getString(Fs.KIND)),
        position = GeoPoint(lat, lng),
        geohash = getString(Fs.GEOHASH).orEmpty(),
        authorId = getString(Fs.AUTHOR_ID).orEmpty(),
        authorName = getString(Fs.AUTHOR_NAME).orEmpty(),
        createdAt = estimatedInstant(Fs.CREATED_AT),
        lastMessageAt = estimatedInstant(Fs.LAST_MESSAGE_AT),
        messageCount = getLong(Fs.MESSAGE_COUNT) ?: 0L,
        lastMediaPath = getString(Fs.LAST_MEDIA_PATH),
        lastMediaKind = getString(Fs.LAST_MEDIA_KIND)?.let { MessageKind.fromId(it) },
    )
}

fun DocumentSnapshot.toMessage(): Message? {
    val kind = MessageKind.fromId(getString(Fs.MESSAGE_KIND))
    val text = getString(Fs.TEXT).orEmpty()
    val imagePath = getString(Fs.IMAGE_PATH)
    val audioPath = getString(Fs.AUDIO_PATH)
    val videoPath = getString(Fs.VIDEO_PATH)
    when (kind) {
        MessageKind.TEXT -> if (text.isEmpty()) return null
        MessageKind.IMAGE -> if (imagePath == null) return null
        MessageKind.VOICE -> if (audioPath == null) return null
        MessageKind.VIDEO -> if (videoPath == null) return null
        MessageKind.STICKER -> if (text.isEmpty()) return null
    }

    val replyId = getString(Fs.REPLY_TO_ID)
    val replyName = getString(Fs.REPLY_TO_AUTHOR_NAME)
    val replyText = getString(Fs.REPLY_TO_TEXT)
    val reply = if (replyId != null && replyName != null && replyText != null) {
        MessageReply(id = replyId, authorName = replyName, text = replyText)
    } else {
        null
    }

    @Suppress("UNCHECKED_CAST")
    val rawReactions = get(Fs.REACTIONS) as? Map<String, Any?> ?: emptyMap()
    val reactions = rawReactions.mapNotNull { (emoji, value) ->
        val uids = when (value) {
            is List<*> -> value.filterIsInstance<String>()
            else -> emptyList()
        }
        if (uids.isEmpty()) null else emoji to uids
    }.toMap()

    return Message(
        id = id,
        kind = kind,
        text = text,
        authorId = getString(Fs.AUTHOR_ID).orEmpty(),
        authorName = getString(Fs.AUTHOR_NAME).orEmpty(),
        authorPhotoURL = getString(Fs.AUTHOR_PHOTO_URL),
        createdAt = estimatedInstant(Fs.CREATED_AT),
        imagePath = imagePath,
        imageWidth = getLong(Fs.IMAGE_WIDTH)?.toInt(),
        imageHeight = getLong(Fs.IMAGE_HEIGHT)?.toInt(),
        audioPath = audioPath,
        audioDurationMs = getLong(Fs.AUDIO_DURATION_MS)?.toInt(),
        videoPath = videoPath,
        videoDurationMs = getLong(Fs.VIDEO_DURATION_MS)?.toInt(),
        videoWidth = getLong(Fs.VIDEO_WIDTH)?.toInt(),
        videoHeight = getLong(Fs.VIDEO_HEIGHT)?.toInt(),
        reply = reply,
        reactions = reactions,
        editedAt = estimatedInstant(Fs.EDITED_AT),
    )
}

private fun DocumentSnapshot.estimatedInstant(field: String): Instant? {
    val timestamp = get(field, Timestamp::class.java, DocumentSnapshot.ServerTimestampBehavior.ESTIMATE)
    return timestamp?.let { Instant.ofEpochSecond(it.seconds, it.nanoseconds.toLong()) }
}
